akka-distributed-data/src/main/scala/akka/cluster/ddata/Flag.scala

```diff
@@ -3,38 +3,78 @@
  */
 package akka.cluster.ddata
 
+import scala.runtime.ScalaRunTime
+
 object Flag {
   /**
    * `Flag` that is initialized to `false`.
    */
   val empty = new Flag(false)
+
+  val Enabled = new Flag(true)
+
   def apply(): Flag = empty
+
+  def apply(enabled: Boolean): Flag = if (enabled) Enabled else empty
+
+  def unapply(flag: Flag): Option[Boolean] = Some(flag.enabled)
+
   /**
    * Java API: `Flag` that is initialized to `false`.
    */
   def create(): Flag = empty
 
-  // unapply from case class
+  /**
+   * Java API: Create a `Flag` given a boolean value.
+   */
+  def create(enabled: Boolean): Flag = Flag(enabled)
+
 }
 
 /**
  * Implements a boolean flag CRDT that is initialized to `false` and
  * can be switched to `true`. `true` wins over `false` in merge.
  *
- * This class is immutable, i.e. "modifying" methods return a new instance.
+ * This class is immutable, i.e. "modifying" methods return either `Flag.empty` or `Flag.Enabled`.
  */
 @SerialVersionUID(1L)
-final case class Flag(enabled: Boolean) extends ReplicatedData with ReplicatedDataSerialization {
+final class Flag private (val enabled: Boolean) extends ReplicatedData with ReplicatedDataSerialization with Product {
 
   type T = Flag
 
-  def switchOn: Flag =
-    if (enabled) this
-    else Flag(true)
+  def switchOn: Flag = Flag.Enabled
 
   override def merge(that: Flag): Flag =
     if (that.enabled) that
     else this
+
+  // reimplementation of all methods the compiler auto-generates for a case class (synthetic methods)
+  // because it was changed from `case class` to a regular class
+  // See issue #24199
+  def copy(enabled: Boolean = this.enabled): Flag = Flag(enabled)
+
+  override def equals(o: Any): Boolean =
+    o match {
+      case other: Flag ⇒
+        this.enabled == other.enabled
+      case _ ⇒ false
+    }
+
+  override def hashCode: Int = ScalaRunTime._hashCode(this)
```

Comment:
used `scalac -Xprint:typer MyCaseClass.scala` to know the generated implementation of `hashCode`